## 휴먼지능정보공학과 201910803 박채희 
### 중간고사 답안

## 문제 1: 열린데이터 수집

서울시 열린데이터에서 검색을 하면 '서울시 우리마을가게 상권분석서비스(상권-아파트)'가 있다.

(https://data.seoul.go.kr/dataList/OA-15566/S/1/datasetView.do?tab=A)

OpenAPI를 통해서 2021년 최근 100개 데이터를 JSON 형식으로 수집하자.

### 1) 프로그램으로 url을 만들어 출력. 그 url을 클릭하면 데이터가 출력 되어야 한다.

In [3]:
import os
from src import mylib 

keyPath=os.path.join(os.getcwd(), 'src', 'key.properties')
key=mylib.getKey(keyPath)
KEY=key['dataseoul'] 

In [25]:
import os
TYPE='json'
SERVICE='InfoTrdarAptQq'
START_INDEX=str(1)
END_INDEX=str(100)
STDR_YM_CD=str(2021)
params=os.path.join(KEY,TYPE,SERVICE,START_INDEX,END_INDEX,STDR_YM_CD)

In [26]:
import urllib.parse

_url='http://openAPI.seoul.go.kr:8088/'
url=urllib.parse.urljoin(_url,params)

#### 답

In [28]:
print(url)

http://openAPI.seoul.go.kr:8088/73725974496a736c34395757705847/json/InfoTrdarAptQq/1/100/2021


### 2) 전체 데이터 중 1건만 출력.




In [29]:
import requests
r=requests.get(url)
trade=r.json()

#### 답)

In [30]:
trade['InfoTrdarAptQq']['row'][0]

{'AE_132_SQMT_HSHLD_CO': 0.0,
 'AE_165_SQMT_HSHLD_CO': 0.0,
 'AE_66_SQMT_BELO_HSHLD_CO': 32.0,
 'AE_66_SQMT_HSHLD_CO': 0.0,
 'AE_99_SQMT_HSHLD_CO': 0.0,
 'APT_HSMP_CO': 5.0,
 'AVRG_AE': 44.0,
 'AVRG_MKTC': 150247619.0,
 'PC_1_HDMIL_BELO_HSHLD_CO': 0.0,
 'PC_1_HDMIL_HSHLD_CO': 32.0,
 'PC_2_HDMIL_HSHLD_CO': 0.0,
 'PC_3_HDMIL_HSHLD_CO': 0.0,
 'PC_4_HDMIL_HSHLD_CO': 0.0,
 'PC_5_HDMIL_HSHLD_CO': 0.0,
 'PC_6_HDMIL_ABOVE_HSHLD_CO': 0.0,
 'STDR_QU_CD': '1',
 'STDR_YY_CD': '2021',
 'TRDAR_CD': '1000975',
 'TRDAR_CD_NM': '동남로82길',
 'TRDAR_SE_CD': 'A',
 'TRDAR_SE_CD_NM': '골목상권'}

### 3) 다음 항목을 100개 정렬해서 출력.

	TRDAR_SE_CD_NM	상권_구분_코드_명

	APT_HSMP_CO	아파트_단지_수

	PC_6_HDMIL_ABOVE_HSHLD_CO	아파트_가격_6_억_이상_세대_수

	AVRG_MKTC	아파트_평균_시가

	TRDAR_CD_NM	상권_코드_명

정렬한 출력예시:

```

골목상권	5.0	0.0	150247619.0	동남로82길

발달상권	3.0	0.0	161157476.0	당산역_2

발달상권	1.0	0.0	81750000.0	서울 성북구 안암역

```

In [33]:
for e in trade['InfoTrdarAptQq']['row']:
    print ("{0}\t{1}\t{2}\t{3}\t{4}\n".format(e['TRDAR_SE_CD_NM'], e['APT_HSMP_CO'], e['PC_6_HDMIL_ABOVE_HSHLD_CO'], e['AVRG_MKTC'], e['TRDAR_CD_NM']))

골목상권	5.0	0.0	150247619.0	동남로82길

발달상권	3.0	0.0	161157476.0	당산역_2

발달상권	1.0	0.0	81750000.0	서울 성북구 안암역

전통시장	37.0	1.0	183067389.0	길동복조리시장

관광특구	71.0	87.0	409473123.0	이태원 관광특구

관광특구	8.0	271.0	237967427.0	동대문패션타운 관광특구

관광특구	5.0	0.0	112097903.0	종로·청계 관광특구

관광특구	22.0	1156.0	527347107.0	잠실 관광특구

골목상권	26.0	22.0	249322039.0	계동길

골목상권	5.0	0.0	87700490.0	난계로27길

골목상권	5.0	0.0	161118305.0	돈화문로11가길

골목상권	41.0	0.0	105457249.0	명륜길

골목상권	41.0	0.0	214023214.0	백석동길

골목상권	3.0	0.0	213246032.0	북촌로11길

골목상권	2.0	2.0	443000000.0	북촌로5길

골목상권	5.0	6.0	430007143.0	북촌로5나길

골목상권	1.0	2.0	572000000.0	삼청로5길

골목상권	12.0	0.0	152368531.0	성균관로5길

골목상권	19.0	0.0	141952840.0	숭인동길

골목상권	100.0	0.0	177741821.0	옥인길

골목상권	1.0	0.0	112279412.0	율곡로16길

골목상권	20.0	0.0	153940955.0	이화장길

골목상권	9.0	0.0	256264683.0	자하문로7길

골목상권	3.0	0.0	81284178.0	종로58가길

골목상권	87.0	0.0	124591950.0	종로65길

골목상권	15.0	0.0	185365873.0	지봉로8길

골목상권	8.0	0.0	142664918.0	창경궁로35길

골목상권	17.0	0.0	107096078.0	창신2길

골목상권	8.0	0.0	95141667.0	창신5길

골목상권	86.0	0.0	116533576.0	창신길

## 문제 2: RDD를 사용해서 단어빈도 계산

다음은 일부 수정된 노래가사이다.

```
yesterday=[

    ["Yesterday 어제는"],

    ["All my troubles seemed to far away 내 모든 문제들이 참 멀리 있는 것 같았는데"],

    ["Now it looks as though they're here to stay 이제는 그 모든 문제들이 여기에 있는 것처럼 보여"],

    ["Oh, I believe in yesterday 오, 어제가 좋았는데"],

    ["Suddenly 갑자기 그-림-자-가 2021 그-림-자-가 2021"],

    ["I'm not half the man I used to be 난 예전 내 모습의 반도 못한 사람이 되어버렸어"],

    ["There's a shadow hanging over me 내 위에 그림자가 드리워져 있네"],

    ["Oh!, yesterday came suddenly 오!, 이제는 너무 갑자기 와버렸어.."],

]
```

In [44]:
import pyspark

In [45]:
myConf=pyspark.SparkConf()
spark=pyspark.sql.SparkSession.builder.master("local").appName("myApp").config(conf=myConf).getOrCreate()

In [162]:
yesterday=[

    ["Yesterday 어제는"],

    ["All my troubles seemed to far away 내 모든 문제들이 참 멀리 있는 것 같았는데"],

    ["Now it looks as though they're here to stay 이제는 그 모든 문제들이 여기에 있는 것처럼 보여"],

    ["Oh, I believe in yesterday 오, 어제가 좋았는데"],

    ["Suddenly 갑자기 그-림-자-가 2021 그-림-자-가 2021"],

    ["I'm not half the man I used to be 난 예전 내 모습의 반도 못한 사람이 되어버렸어"],

    ["There's a shadow hanging over me 내 위에 그림자가 드리워져 있네"],

    ["Oh!, yesterday came suddenly 오!, 이제는 너무 갑자기 와버렸어.."],

]

### 1) 단어빈도를 계산해서 출력

출력 예시: [('Yesterday', 1),('어제는', 1),('All', 1),('my', 1),...]

In [165]:
myRdd=spark.sparkContext.parallelize(yesterday)

#### 답)

In [188]:
wc1=myRdd\
    .flatMap(lambda x:x)\
    .flatMap(lambda x:x.split())\
    .map(lambda x:(x,1))\
    .groupByKey()\
    .mapValues(sum)\
    .take(100)
print(wc1)

[('Yesterday', 1), ('어제는', 1), ('All', 1), ('my', 1), ('troubles', 1), ('seemed', 1), ('to', 3), ('far', 1), ('away', 1), ('내', 3), ('모든', 2), ('문제들이', 2), ('참', 1), ('멀리', 1), ('있는', 2), ('것', 1), ('같았는데', 1), ('Now', 1), ('it', 1), ('looks', 1), ('as', 1), ('though', 1), ("they're", 1), ('here', 1), ('stay', 1), ('이제는', 2), ('그', 1), ('여기에', 1), ('것처럼', 1), ('보여', 1), ('Oh,', 1), ('I', 2), ('believe', 1), ('in', 1), ('yesterday', 2), ('오,', 1), ('어제가', 1), ('좋았는데', 1), ('Suddenly', 1), ('갑자기', 2), ('그-림-자-가', 2), ('2021', 2), ("I'm", 1), ('not', 1), ('half', 1), ('the', 1), ('man', 1), ('used', 1), ('be', 1), ('난', 1), ('예전', 1), ('모습의', 1), ('반도', 1), ('못한', 1), ('사람이', 1), ('되어버렸어', 1), ("There's", 1), ('a', 1), ('shadow', 1), ('hanging', 1), ('over', 1), ('me', 1), ('위에', 1), ('그림자가', 1), ('드리워져', 1), ('있네', 1), ('Oh!,', 1), ('came', 1), ('suddenly', 1), ('오!,', 1), ('너무', 1), ('와버렸어..', 1)]


### 2) 빈도순으로 출력, 빈도 1이하는 제외.

In [189]:
wc2=myRdd\
    .flatMap(lambda x:x)\
    .flatMap(lambda x:x.split())\
    .map(lambda x:(x,1))\
    .groupByKey()\
    .mapValues(sum)\
    .map(lambda x:(x[1],x[0]))\
    .sortByKey(False)\
    .take(100)

#### 답)

In [203]:
for i in wc2:
  if i[0]>1:
    print ('단어:',i[1], ' / 빈도:',i[0])

단어: to  / 빈도: 3
단어: 내  / 빈도: 3
단어: 모든  / 빈도: 2
단어: 문제들이  / 빈도: 2
단어: 있는  / 빈도: 2
단어: 이제는  / 빈도: 2
단어: I  / 빈도: 2
단어: yesterday  / 빈도: 2
단어: 갑자기  / 빈도: 2
단어: 그-림-자-가  / 빈도: 2
단어: 2021  / 빈도: 2


### 3) 소문자, 하이픈, 느낌표, 숫자를 제외하고 단어빈도 계산해서 출력 (빈도 1이하는 제외)

* 소문자: 'Yesterday', 'yesterday'를 동일한 단어로 계산

* 하이픈: '그림자가', '그-림-자-가'를 동일한 단어로 계산

* 느낌표: 'oh', 'oh!'를 동일한 단어로 계산

* 숫자: '2021'는 제외하고 계산

In [264]:
wc3=myRdd\
    .flatMap(lambda x:x)\
    .flatMap(lambda x:x.split())\
    .map(lambda x: x.replace('!','').replace('-','').lower())\
    .filter(lambda x: x!='2021')\
    .map(lambda x:(x,1))\
    .reduceByKey(lambda x,y:x+y)\
    .map(lambda x:(x[1],x[0]))\
    .take(100)
print(wc3)

[(3, 'yesterday'), (1, '어제는'), (1, 'all'), (1, 'my'), (1, 'troubles'), (1, 'seemed'), (3, 'to'), (1, 'far'), (1, 'away'), (3, '내'), (2, '모든'), (2, '문제들이'), (1, '참'), (1, '멀리'), (2, '있는'), (1, '것'), (1, '같았는데'), (1, 'now'), (1, 'it'), (1, 'looks'), (1, 'as'), (1, 'though'), (1, "they're"), (1, 'here'), (1, 'stay'), (2, '이제는'), (1, '그'), (1, '여기에'), (1, '것처럼'), (1, '보여'), (2, 'oh,'), (2, 'i'), (1, 'believe'), (1, 'in'), (2, '오,'), (1, '어제가'), (1, '좋았는데'), (2, 'suddenly'), (2, '갑자기'), (3, '그림자가'), (1, "i'm"), (1, 'not'), (1, 'half'), (1, 'the'), (1, 'man'), (1, 'used'), (1, 'be'), (1, '난'), (1, '예전'), (1, '모습의'), (1, '반도'), (1, '못한'), (1, '사람이'), (1, '되어버렸어'), (1, "there's"), (1, 'a'), (1, 'shadow'), (1, 'hanging'), (1, 'over'), (1, 'me'), (1, '위에'), (1, '드리워져'), (1, '있네'), (1, 'came'), (1, '너무'), (1, '와버렸어..')]


#### 답)

In [265]:
for i in wc3:
  if i[0]>1:
    print ('단어:',i[1], ' / 빈도:',i[0])

단어: yesterday  / 빈도: 3
단어: to  / 빈도: 3
단어: 내  / 빈도: 3
단어: 모든  / 빈도: 2
단어: 문제들이  / 빈도: 2
단어: 있는  / 빈도: 2
단어: 이제는  / 빈도: 2
단어: oh,  / 빈도: 2
단어: i  / 빈도: 2
단어: 오,  / 빈도: 2
단어: suddenly  / 빈도: 2
단어: 갑자기  / 빈도: 2
단어: 그림자가  / 빈도: 3


### 4) 1글자 단어 (예를 들어 'i', '내') 제외하고 단어빈도 계산 (빈도 1이하는 제외)

#### 답)

In [268]:
for i in wc3:
  if i[0]>1:
    if len(i[1])>1:
      print ('단어:',i[1], ' / 빈도:',i[0])

단어: yesterday  / 빈도: 3
단어: to  / 빈도: 3
단어: 모든  / 빈도: 2
단어: 문제들이  / 빈도: 2
단어: 있는  / 빈도: 2
단어: 이제는  / 빈도: 2
단어: oh,  / 빈도: 2
단어: 오,  / 빈도: 2
단어: suddenly  / 빈도: 2
단어: 갑자기  / 빈도: 2
단어: 그림자가  / 빈도: 3


추가적으로 '오,' 같은 단어들도 ','를 제외하여
(```replace (',','') ```) 빈도를 출력하면 '오' 단어도 1글자 단어가 되어 단어빈도 계산에서 제외될 것 이다.

## 문제 3: DataFrame을 사용해서 코로나 발생동향 분석

서울시 열린데이터에서 검색을 하면 '서울특별시 코로나19 자치구별 확진자 발생동향'이 있다.

https://data.seoul.go.kr/dataList/OA-20470/S/1/datasetView.do

데이터를 '내려받기'해서 '서울특별시\ 코로나19\ 자치구별\ 확진자\ 발생동향.json' 파일을 분석하세요.






3) (null제거하고) 종로구의 2020년 3월 데이터 출력.

+-------------+------+

|       jcg_dt|jongno|

+-------------+------+

|2020.03.31.10|    13|

...



4) (null제거하고) 년 x 월별 합계 10건 출력.

+-----+-------+-------+

|month|   2020|   2021|

+-----+-------+-------+

|   07| 1273.0|62849.0|

...




### 1) 전체 건수 출력. null 값을 제외한 null 건수 출력 (null이 남아있는지 확인하기 위해)

In [417]:
covid=spark.read.json('서울특별시 코로나19 자치구별 확진자 발생동향.json')

In [418]:
covid.show()

+--------------------+----+------+------+---------+-------+----------+----+-----+----+------+-------+----------+--------+-----------+-------+----------+-------+----------+---------+------------+----+-------+------+---------+--------+-----------+-------------+------+---------+------+---------+--------+-----------+----+-------+-----+--------+----+------+------+---------+--------+-----------+---------+------------+------+---------+-------------+---------+------------+----+------+-------+----------+
|     _corrupt_record| ddm|ddmadd|dobong|dobongadd|dongjak|dongjakadd|  ep|epadd| etc|etcadd|gangbuk|gangbukadd|gangdong|gangdongadd|gangnam|gangnamadd|gangseo|gangseoadd|geumcheon|geumcheonadd|guro|guroadd|gwanak|gwanakadd|gwangjin|gwangjinadd|       jcg_dt|jongno|jongnoadd|junggu|jungguadd|jungnang|jungnangadd|mapo|mapoadd|nowon|nowonadd| sdm|sdmadd|seocho|seochoadd|seongbuk|seongbukadd|seongdong|seongdongadd|songpa|songpaadd|    work_dttm|yangcheon|yangcheonadd| ydp|ydpadd|yongsan|yongsanadd

In [410]:
covid.dropna()

DataFrame[_corrupt_record: string, ddm: string, ddmadd: string, dobong: string, dobongadd: string, dongjak: string, dongjakadd: string, ep: string, epadd: string, etc: string, etcadd: string, gangbuk: string, gangbukadd: string, gangdong: string, gangdongadd: string, gangnam: string, gangnamadd: string, gangseo: string, gangseoadd: string, geumcheon: string, geumcheonadd: string, guro: string, guroadd: string, gwanak: string, gwanakadd: string, gwangjin: string, gwangjinadd: string, jcg_dt: string, jongno: string, jongnoadd: string, junggu: string, jungguadd: string, jungnang: string, jungnangadd: string, mapo: string, mapoadd: string, nowon: string, nowonadd: string, sdm: string, sdmadd: string, seocho: string, seochoadd: string, seongbuk: string, seongbukadd: string, seongdong: string, seongdongadd: string, songpa: string, songpaadd: string, work_dttm: bigint, yangcheon: string, yangcheonadd: string, ydp: string, ydpadd: string, yongsan: string, yongsanadd: string]

In [411]:
covid.show()

+--------------------+----+------+------+---------+-------+----------+----+-----+----+------+-------+----------+--------+-----------+-------+----------+-------+----------+---------+------------+----+-------+------+---------+--------+-----------+-------------+------+---------+------+---------+--------+-----------+----+-------+-----+--------+----+------+------+---------+--------+-----------+---------+------------+------+---------+-------------+---------+------------+----+------+-------+----------+
|     _corrupt_record| ddm|ddmadd|dobong|dobongadd|dongjak|dongjakadd|  ep|epadd| etc|etcadd|gangbuk|gangbukadd|gangdong|gangdongadd|gangnam|gangnamadd|gangseo|gangseoadd|geumcheon|geumcheonadd|guro|guroadd|gwanak|gwanakadd|gwangjin|gwangjinadd|       jcg_dt|jongno|jongnoadd|junggu|jungguadd|jungnang|jungnangadd|mapo|mapoadd|nowon|nowonadd| sdm|sdmadd|seocho|seochoadd|seongbuk|seongbukadd|seongdong|seongdongadd|songpa|songpaadd|    work_dttm|yangcheon|yangcheonadd| ydp|ydpadd|yongsan|yongsanadd

In [424]:
import pandas as pd
covid=covid.toPandas() 

In [425]:
covid=covid[3:]

In [426]:
len(covid)

630

#### 답) 총 630건의 데이터입니다.



### 2) (null제거하고) 종로, 마포, 용산, 합계 (종로, 마포, 용산) 10건 출력.

+------+----+-------+------+

|jongno|mapo|yongsan|   tot|

+------+----+-------+------+

|  1968|4127|   3031|9126.0|

뒤의 문제는 시간이 없어서 못풀었습니다.